### PR TITLE
Added support for vector envs  in evaluation

### LINKED ISSUE #402: [Feature Request] support vec env evaluation with n_envs >1
Description

Issue for #398

Right now, policy evaluation can only happen on vector environments with n=1 copies of the env. This is a small change to expand this.

Motivation and Context

-When doing hyperparameter tuning on certain environments, large numbers of evaluations are required to get accurate performance metrics. This lets that happen much faster.
-This is required for all the wrappers to work when PettingZoo environments are used

Both of these have been problems for be recently.

This change is so small that I don't know what tests if any would be sensible to add.

### Alternatives

Having to manage my own fork with this functionality

### Additional context

Similar libraries do this:

https://github.com/cpnota/autonomous-learning-library/blob/develop/all/experiments/parallel_env_experiment.py#L90

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by benblack769**
## Description

Modifies EvalCallback and evaluate_policy in order to allow for vector envs with num_envs > 1. 

To solve bias towards short episodes mentioned in #402, each element of the vector env is assigned the number of episodes it needs to evaluate ahead of time, and any additional episodes a sub env evaluates are ignored. 

Modified the `test_evaluate_policy_monitors` test to make it create vector envs with > 1 envs. Let me know if other tests are needed. 

## Motivation and Context

closes #402 

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

### CODE REVIEW

File: stable_baselines3/common/callbacks.py [hunk 1]
```diff
@@ -342,9 +339,6 @@ def _log_success_callback(self, locals_: Dict[str, Any], globals_: Dict[str, Any
         :param globals_:
         """
         info = locals_["info"]
-        # VecEnv: unpack
```
Comment (araffin) [inline stable_baselines3/common/callbacks.py:345]: the code below should be updated too (counting successes)

File: stable_baselines3/common/evaluation.py [hunk 2]
```diff
@@ -52,14 +52,13 @@ def evaluate_policy(
...
-    from stable_baselines3.common.env_util import is_wrapped
     from stable_baselines3.common.monitor import Monitor
 
-    if isinstance(env, VecEnv):
-        assert env.num_envs == 1, "You must pass only one environment when using this function"
-        is_monitor_wrapped = is_vecenv_wrapped(env, VecMonitor) or env.env_is_wrapped(Monitor)[0]
-    else:
-        is_monitor_wrapped = is_wrapped(env, Monitor)
+    if not isinstance(env, VecEnv):
+        gym_env = env
```
Comment (araffin) [inline stable_baselines3/common/evaluation.py:58]: why not `env = DummyVecEnv([lambda: env])` directly?

also please add a comment of why.

File: stable_baselines3/common/evaluation.py [hunk 3]
```diff
@@ -69,41 +68,52 @@ def evaluate_policy(
...
+    while (episode_counts < episode_count_targets).any():
+        actions, states = model.predict(obss, state=states, deterministic=deterministic)
+        obss, rewards, dones, infos = env.step(actions)
+        current_rewards += rewards
+        current_lengths += 1
+        for i in range(n_envs):
+            if episode_counts[i] < episode_count_targets[i]:
+                reward = rewards[i]
+                done = dones[i]
+                # obs, reward, done = obss[i], rewards[i], dones[i]
```
Comment (araffin) [inline stable_baselines3/common/evaluation.py:91]: please remove debug code

File: stable_baselines3/common/evaluation.py [hunk 4]
```diff
@@ -21,7 +21,8 @@ def evaluate_policy(
 ) -> Union[Tuple[float, float], Tuple[List[float], List[int]]]:
     """
     Runs policy for ``n_eval_episodes`` episodes and returns average reward.
-    This is made to work only with one env.
+    If a vector env is passed in, this divides the episodes to evaluate onto the
+    different elements of the vector env.
```
Comment (araffin) [inline stable_baselines3/common/evaluation.py:25]: Please put a link to the GH issue here.

File: stable_baselines3/common/evaluation.py [hunk 5]
```diff
@@ -69,41 +68,52 @@ def evaluate_policy(
...
-                callback(locals(), globals())
-            episode_length += 1
-            if render:
-                env.render()
+    n_envs = env.num_envs
+    episode_rewards = []
+    episode_lengths = []
+
+    episode_counts = np.zeros(n_envs, dtype="int")
+    episode_count_targets = np.array([(n_eval_episodes + i) // n_envs for i in range(n_envs)], dtype="int")
```
Comment (araffin) [inline stable_baselines3/common/evaluation.py:76]: shouldn't the episode count target be the same for each env?

File: stable_baselines3/common/evaluation.py [hunk 6]
```diff
@@ -69,41 +68,52 @@ def evaluate_policy(
...
+
+    current_rewards = np.zeros(n_envs)
+    current_lengths = np.zeros(n_envs, dtype="int")
+    obss = env.reset()
```
Comment (araffin) [inline stable_baselines3/common/evaluation.py:80]: Please use `observations` instead

File: tests/test_utils.py [hunk 7]
```diff
@@ -196,7 +196,8 @@ def reset(self, **kwargs):
 def test_evaluate_policy_monitors(vec_env_class):
     # Test that results are correct with monitor environments.
     # Also test VecEnvs
-    n_eval_episodes = 2
+    n_eval_episodes = 3
```
Comment (araffin) [inline tests/test_utils.py:199]: Please also add a test that check that the number of episodes is the one expected and make `n_envs` a parameter of the test function.
You should also add a test for the evaluation callback with multiple envs.

File: stable_baselines3/common/callbacks.py (hunk 1 above)
Comment (benblack769) [reply stable_baselines3/common/callbacks.py:358]: I think this code works as is. The evaluation code unpacks the results of the vector env so that the callback can count scalar values. For example the test on line 105 of test_callbacks.py passes.

File: stable_baselines3/common/evaluation.py (hunk 2 above)
Comment (benblack769) [reply stable_baselines3/common/evaluation.py]: I guess it works to do it that way. I was just unnecessarily nervous because using lambdas without care has screwed me over before. Changed.

File: stable_baselines3/common/evaluation.py (hunk 3 above)
Comment (benblack769) [reply stable_baselines3/common/evaluation.py]: Sorry!

File: stable_baselines3/common/evaluation.py (hunk 5 above)
Comment (benblack769) [reply stable_baselines3/common/evaluation.py:75]: This is to handle odd numbers of episodes. For example if n_envs is 2 and and n_eval_episodes is 3 , then one environment will evaluate 2 episodes and the other will evaluate 1. This is just a formula that distributes them.

File: tests/test_utils.py (hunk 7 above)
Comment (benblack769) [reply tests/test_utils.py:229]: I added two tests for these cases.

File: stable_baselines3/common/evaluation.py (hunk 6 above)
Comment (benblack769) [reply stable_baselines3/common/evaluation.py]: Changed

**Comment by benblack769** [comment]
@araffin I think I fixed all the issues you pointed out. Please give it another look.

File: stable_baselines3/common/callbacks.py (hunk 1 above)
Comment (araffin) [reply stable_baselines3/common/callbacks.py:358]: alright, I did not realize that you were unpacking later

File: stable_baselines3/common/evaluation.py (hunk 5 above)
Comment (araffin) [reply stable_baselines3/common/evaluation.py:78]: thanks for the clarification and good that you added a comment ;)

File: stable_baselines3/common/evaluation.py [hunk 8]
```diff
@@ -69,41 +69,53 @@ def evaluate_policy(
...
+                # unpack values so that the callback can access the local variables
+                reward = rewards[i]
+                done = dones[i]
+                info = infos[i]
+
+                if callback is not None:
+                    callback(locals(), globals())
+
+                if dones[i]:
+                    if is_monitor_wrapped:
```
Comment (araffin) [inline stable_baselines3/common/evaluation.py:101]: minor: the two if can probably be combined, no?

File: tests/test_utils.py [hunk 9]
```diff
@@ -197,11 +197,37 @@ def reset(self, **kwargs):
...
 
+@pytest.mark.parametrize("n_envs", [1, 2, 5, 7])
+def test_evaluate_vector_env(n_envs):
+    # Tests that the number of episodes evaluated is correct
+    n_eval_episodes = 6
+
+    env_id = "CartPole-v0"
+    env = gym.make(env_id)
+    model = A2C("MlpPolicy", env_id, seed=0)
+    # model.learn(100000)
```
Comment (araffin) [inline tests/test_utils.py:208]: minor: please remove commented code

File: tests/test_utils.py [hunk 10]
```diff
@@ -197,11 +197,37 @@ def reset(self, **kwargs):
         return self.last_obs
 
 
+@pytest.mark.parametrize("n_envs", [1, 2, 5, 7])
+def test_evaluate_vector_env(n_envs):
```
Comment (araffin) [inline tests/test_utils.py:201]: `n_envs` is not used...

File: docs/misc/changelog.rst [hunk 11]
```diff
@@ -45,6 +45,7 @@ New Features:
 - Added support for image observation when using ``HER``
 - Added ``replay_buffer_class`` and ``replay_buffer_kwargs`` arguments to off-policy algorithms
 - Added ``kl_divergence`` helper for ``Distribution`` classes (@09tangriro)
+- Added support for vector environments with num_envs > 1 (@benblack769)
```
Comment (araffin) [inline docs/misc/changelog.rst:48]: please add your username at the bottom too ;)

**Review by araffin — COMMENTED** [review]
only minor remarks, other LGTM ;)

File: stable_baselines3/common/evaluation.py (hunk 8 above)
Comment (benblack769) [reply stable_baselines3/common/evaluation.py:101]: No. Everything needs to be kept seperate. In particular, if `is_monitor_wrapped` is True but not  `"episode" in info.keys()`, then it should not add to the episode_rewards list.

File: tests/test_utils.py (hunk 9 above)
Comment (benblack769) [reply tests/test_utils.py]: Removed.

File: tests/test_utils.py (hunk 10 above)
Comment (benblack769) [reply tests/test_utils.py:201]: Fixed.

File: docs/misc/changelog.rst (hunk 11 above)
Comment (benblack769) [reply docs/misc/changelog.rst:48]: Done

**Comment by benblack769** [comment]
Ok, I think I addressed these final issues.

File: stable_baselines3/common/evaluation.py (hunk 8 above)
Comment (araffin) [reply stable_baselines3/common/evaluation.py:101]: I meant something like `if is_monitor_wrapped and "episode" in info.keys():`

**Review by araffin — APPROVED** [review]
LGTM, thanks =)

File: stable_baselines3/common/evaluation.py (hunk 8 above)
Comment (araffin) [reply stable_baselines3/common/evaluation.py]: nevermind...

**Comment by araffin** [comment]
One last thing I'm currently investigating, `tests/test_utils.py::test_evaluate_policy_monitors` is throwing warnings (even with `n_envs=1`) where as it was not the case previously...

**Comment by araffin** [comment]
> One last thing I'm currently investigating, `tests/test_utils.py::test_evaluate_policy_monitors` is throwing warnings (even with `n_envs=1`) where as it was not the case previously...

fixed, the count of episodes was wrong in some edge case...

**Comment by benblack769** [comment]
Thanks for cleaning up the remaining issues.
